### PR TITLE
Log ntpd startup failures

### PR DESCRIPTION
**Pull request by eriknordmark**
We've seen one case of a watchdog due to ntpd.pid not running soon after boot. In fact it looks like it never started. Adding logging of exit codes to we can see if/when it happens again.

### CODE REVIEW

File: pkg/pillar/scripts/device-steps.sh [hunk 1]
```diff
@@ -295,8 +295,8 @@ access_usb() {
     if [ -n "$SPECIAL" ] && [ -b "$SPECIAL" ]; then
         echo "$(date -Ins -u) Found USB with DevicePortConfig: $SPECIAL"
         if ! mount -t vfat "$SPECIAL" /mnt; then
-            # XXX !? will be zero
-            echo "$(date -Ins -u) mount $SPECIAL failed: $?"
+            ret_code=$?
```
Comment (giggsoff) [inline pkg/pillar/scripts/device-steps.sh:298]: Cannot understand code from what command we expect here.

File: pkg/pillar/scripts/device-steps.sh [hunk 2]
```diff
@@ -578,7 +587,9 @@ while true; do
         echo "$(date -Ins -u) NTP server changed from $NTPSERVER to $ns"
         NTPSERVER="$ns"
         kill "$(cat /run/ntpd.pid)"
+        echo "$(date -Ins -u) ntpd -q -n -g -p $NTPSERVER"
         /usr/sbin/ntpd -p "$NTPSERVER"
+        echo "$(date -Ins -u) ntpd: $?"
```
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh:592]: Thanks - must have missed that one.

File: pkg/pillar/scripts/device-steps.sh (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh:298]: Intent is to get exit code from the mount. Perhaps that doesn't work in the context of the if statement? (I don't have an easy way to test this - but it isn't worse than the old code getting the exit code from the data command.)

Comment (giggsoff) [reply pkg/pillar/scripts/device-steps.sh:298]: Probably it is better to move the command from the if and check for $ret_code -ne 0 inside if?

File: pkg/pillar/scripts/device-steps.sh [hunk 3]
```diff
@@ -578,7 +587,10 @@ while true; do
         echo "$(date -Ins -u) NTP server changed from $NTPSERVER to $ns"
         NTPSERVER="$ns"
         kill "$(cat /run/ntpd.pid)"
+        echo "$(date -Ins -u) ntpd -q -n -g -p $NTPSERVER"
```
Comment (giggsoff) [inline pkg/pillar/scripts/device-steps.sh:590]: Command and output mismatch for some reason here and few lines below

File: pkg/pillar/scripts/device-steps.sh (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh]: Fixed

File: pkg/pillar/scripts/device-steps.sh (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh]: Oops. Fixed to also pass -g to allow the initial jump for the case where pool.ntp.org didn't work.

**Review by giggsoff — APPROVED** [review]
LGTM

**Review by milan-zededa — APPROVED** [review]

**Comment by milan-zededa** [comment]
Later we might add something like:
```
logcmd() {
	echo $(date -Ins -u) "$@"
	"$@"
	ret_code=$?
        echo "$(date -Ins -u) $1: $ret_code"
}
```
if there are more commands that can hang/fail.

**Comment by gkodali-zededa** [comment]
https://www.cyberciti.biz/faq/linux-run-a-command-with-a-time-limit/

Can the timeout command help here? (I have not used this before and will need some experimenting with it) Especially with the --foreground & --preserve-status options.
We can try starting ntpd for a certain number of iterations. When that fails, we can dump a bunch of system information and exit gracefully. This the failure path is also in our control.

**Comment by eriknordmark** [comment]
@gkodali-zededa @milan-zededa issue I've seen isn't a hang but watchdog soon after boot realizing that ndpd isn't running. 
If we get these logs we might have a clue why. It could be that the restart (kill+start) isn't reliable but that is speculation.

**Pull request merged**